### PR TITLE
updated R2 to include momentum

### PR DESCRIPTION
**Pull request by farhadrclass**
I added the momentum to R2, 
I ran the unit tests, however I am not sure if we need to create a new one just for momentum ?

### CODE REVIEW

**Comment by farhadrclass** [comment]
@dpo  wait on this, I found a small bug with bigfloat

File: src/R2.jl [hunk 1]
```diff
@@ -16,6 +16,7 @@ export R2, R2Solver
 - σmin = eps(T): step parameter for R2 algorithm
 - max_eval::Int: maximum number of evaluation of the objective function
 - max_time::Float64 = 3600.0: maximum time limit in seconds
+- β = T(0.0) is the constact for the Momentum algorithm, it needs to be ∈ [0,1], if 0, R2 does not use Momentum
```
Comment (dpo) [inline src/R2.jl:19]: typo

File: src/R2.jl [hunk 2]
```diff
@@ -85,6 +86,7 @@ function solve!(
   σmin = zero(T),
   max_time::Float64 = 3600.0,
   max_eval::Int = -1,
+  β::T = T(0.0),
```
Comment (dpo) [inline src/R2.jl:89]:
```suggestion
  β::T = T(0),
```

File: src/R2.jl [hunk 3]
```diff
@@ -96,6 +98,8 @@ function solve!(
   x = solver.x .= x0
   ∇fk = solver.gx
   ck = solver.cx
+  # used for Momentum, start with zeros
+  d =  zero(similar(nlp.meta.x0))
```
Comment (dpo) [inline src/R2.jl:102]: should be in solver object

File: src/R2.jl [hunk 4]
```diff
@@ -136,7 +140,8 @@ function solve!(
     (output.status == :user)
 
   while !done
-    ck .= x .- (∇fk ./ σk)
+    d .= ∇fk .* (T(1) - β) + d .* β
```
Comment (dpo) [inline src/R2.jl:143]:
```suggestion
    d .= ∇fk .* (T(1) - β) .+ d .* β  
```

Comment (dpo) [reply src/R2.jl:143]: Maybe you could check if β == 0.

**Review by dpo — CHANGES_REQUESTED** [review]
Thanks. It would be nice to see some benchmarks (performance and allocations).

File: src/R2.jl (hunk 4 above)
Comment (farhadrclass) [reply src/R2.jl:143]: If b is zero then d .= ∇fk

Comment (dpo) [reply src/R2.jl]: Yes. It would be cheaper to compute that way.

File: src/R2.jl [hunk 5]
```diff
@@ -56,13 +56,15 @@ mutable struct R2Solver{T, V}
...
   cx::V
+  d::V
   output::GenericExecutionStats{T, V}
 end
 
 function R2Solver(nlp::AbstractNLPModel{T, V}) where {T, V}
   x = similar(nlp.meta.x0)
   gx = similar(nlp.meta.x0)
   cx = similar(nlp.meta.x0)
+  d = zero(similar(nlp.meta.x0))
```
Comment (dpo) [inline src/R2.jl:67]:
```suggestion
  d = fill!(0, similar(nlp.meta.x0))
```

**Review by dpo — CHANGES_REQUESTED** [review]

File: src/R2.jl (hunk 4 above)
Comment (farhadrclass) [reply src/R2.jl]: I suggest this then 
    if β == 0
      ck .= x .- (∇fk ./ σk)
    else
      d .= ∇fk .* (T(1) - β) .+ d .* β
      ck .= x .- (d ./ σk)
    end

File: src/R2.jl (hunk 5 above)
Comment (farhadrclass) [reply src/R2.jl:67]: This gives me an error and I also tried 
fill!(T(0), similar(nlp.meta.x0))

Comment (farhadrclass) [reply src/R2.jl:67]: Test callback: Error During Test at C:\Users\Farhad\Documents\GitHub\JSOSolvers.jl\test\callback.jl:2
  Got exception outside of a @test
  MethodError: no method matching fill!(::Float64, ::Vector{Float64})
  Closest candidates are:
    fill!(::Array{T, N} where N, ::Any) where T at array.jl:332
    fill!(::BitArray, ::Any) at bitarray.jl:377
    fill!(::SubArray{Bool, var"#s832", var"#s831", Tuple{AbstractUnitRange{Int64}}, L} where {var"#s832", var"#s831"<:BitArray, L}, ::Any) at multidimensional.jl:1421
    ...
  Stacktrace:
    [1] R2Solver(nlp::ADNLPModel{Float64, Vector{Float64}, Vector{Int64}})
      @ JSOSolvers C:\Users\Farhad\Documents\GitHub\JSOSolvers.jl\src\R2.jl:67
    [2] R2(nlp::ADNLPModel{Float64, Vector{Float64}, Vector{Int64}}; kwargs::Base.Iterators.Pairs{Symbol, var"#cb#3"{Vector{Float64}, Vector{Float64}}, Tuple{Symbol}, NamedTuple{(:callback,), Tuple{var"#cb#3"{Vector{Float64}, Vector{Float64}}}}})
      @ JSOSolvers C:\Users\Farhad\Documents\GitHub\JSOSolvers.jl\src\R2.jl:73
    [3] (::var"#1#4"{var"#cb#3"{Vector{Float64}, Vector{Float64}}, ADNLPModel{Float64, Vector{Float64}, Vector{Int64}}})()
      @ Main C:\Users\Farhad\Documents\GitHub\JSOSolvers.jl\test\callback.jl:16
    [4] with_logstate(f::Function, logstate::Any)
      @ Base.CoreLogging .\logging.jl:491
    [5] with_logger(f::Function, logger::Base.CoreLogging.NullLogger)
      @ Base.CoreLogging .\logging.jl:603
    [6] macro expansion
      @ C:\Users\Farhad\Documents\GitHub\JSOSolvers.jl\test\callback.jl:15 [inlined]
    [7] macro expansion
      @ C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\Test\src\Test.jl:1151 [inlined]
    [8] top-level scope
      @ C:\Users\Farhad\Documents\GitHub\JSOSolvers.jl\test\callback.jl:3
    [9] include(fname::String)
      @ Base.MainInclude .\client.jl:444
   [10] top-level scope
      @ C:\Users\Farhad\Documents\GitHub\JSOSolvers.jl\test\runtests.jl:10
   [11] include(fname::String)
      @ Base.MainInclude .\client.jl:444
   [12] top-level scope
      @ none:6
   [13] eval
      @ .\boot.jl:360 [inlined]
   [14] exec_options(opts::Base.JLOptions)
      @ Base .\client.jl:261
   [15] _start()
      @ Base .\client.jl:485

**Comment by farhadrclass** [comment]
> 

I create a new repo for all my benchmarks and I put the benchmark for this there:
https://github.com/farhadrclass/Benchmark/tree/main/R2

Still running but I will update it with Results

File: src/R2.jl (hunk 5 above)
Comment (dpo) [reply src/R2.jl:67]: sorry, it should be `fill!(similar(...), 0)`.

File: src/R2.jl (hunk 4 above)
Comment (dpo) [reply src/R2.jl]: yes, looks good

**Comment by farhadrclass** [comment]
@dpo  it passed every test and ready but I am working on Benchmark now and should have results soon if we can wait for it

File: src/R2.jl [hunk 6]
```diff
@@ -22,6 +22,7 @@ For advanced usage, first define a `R2Solver` to preallocate the memory used in
 - σmin = eps(T): step parameter for R2 algorithm
 - max_eval::Int: maximum number of evaluation of the objective function
 - max_time::Float64 = 3600.0: maximum time limit in seconds
+- β = T(0) is the constant for the Momentum algorithm, it needs to be ∈ [0,1], if 0, R2 does not use Momentum
```
Comment (dpo) [inline src/R2.jl:25]:
```suggestion
- β = T(0) ∈ [0,1] is the constant in the momentum term. If `β == 0`, R2 does not use momentum
```

File: src/R2.jl [hunk 7]
```diff
@@ -109,6 +113,8 @@ function SolverCore.solve!(
   x = solver.x .= x0
   ∇fk = solver.gx
   ck = solver.cx
+  # used for Momentum, start with zeros
```
Comment (dpo) [inline src/R2.jl:116]:
```suggestion
```

File: src/R2.jl [hunk 8]
```diff
@@ -69,13 +70,15 @@ mutable struct R2Solver{T, V} <: AbstractOptimizationSolver
   x::V
   gx::V
   cx::V
+  d::V
```
Comment (dpo) [inline src/R2.jl:73]:
```suggestion
  d::V   # used for momentum term
```

**Review by dpo — CHANGES_REQUESTED** [review]
Thank you. Almost there.

File: src/R2.jl [hunk 9]
```diff
@@ -98,17 +101,19 @@ function SolverCore.solve!(
   σmin = zero(T),
   max_time::Float64 = 3600.0,
   max_eval::Int = -1,
+  β = T(0),
   verbose::Int = 0,
 ) where {T, V}
   unconstrained(nlp) || error("R2 should only be called on unconstrained problems.")
-
+  β = T(β)
```
Comment (dpo) [inline src/R2.jl:108]:
```suggestion
```

**Review by dpo — CHANGES_REQUESTED** [review]

File: src/R2.jl (hunk 9 above)
Comment (dpo) [reply src/R2.jl]: Note that there are two different β symbols here.

Comment (farhadrclass) [reply src/R2.jl]: yes, I removed it since it was a typo

**Pull request merged**